BUG: to_clipboard fails to format output for Excel

david-liu-brattle-1 opened the pull request:
`DataFrame.to_clipboard` has been broken for pasting to excel.  Tables are copied with spaces as delimiters instead of tabs (#21104).  

This issue originated in https://github.com/pandas-dev/pandas/commit/e1d5a2738235fec22f3cfad4814e09e3e3786f8c#diff-3f25860d9237143c1952a1f93c3aae18R102
which I've partially reverted.  

By setting the delimiter to `r'\t'`, a 2 character string, obj.to_csv raised an error, but is was caught and passed silently.  I reverted the separator to `'\t'`. 

Similar issue in `read_clipboard` also fixed.


- [x] closes #21104

david-liu-brattle-1 committed:
Fixed copy table to excel

Reverted a change in e1d5a2738235fec22f3cfad4814e09e3e3786f8c

WillAyd commented:
Can you add a test to make sure this produces the correct output?

WillAyd reviewed (changes requested):
Can you add a test to make sure this produces the correct output? Also could use a whatsnew note

david-liu-brattle-1 committed:
Unit Test and whatsnew

pep8speaks commented:
Hello @david-liu-brattle-1! Thanks for updating the PR.



 - In the file [`pandas/tests/io/test_clipboard.py`](https://github.com/pandas-dev/pandas/blob/a8c098d596c39df7f4599bf1da516d00be7de226/pandas/tests/io/test_clipboard.py), following are the PEP8 issues :

> [Line 127:5](https://github.com/pandas-dev/pandas/blob/a8c098d596c39df7f4599bf1da516d00be7de226/pandas/tests/io/test_clipboard.py#L127): [E301](https://duckduckgo.com/?q=pep8%20E301) expected 1 blank line, found 0
> [Line 129:29](https://github.com/pandas-dev/pandas/blob/a8c098d596c39df7f4599bf1da516d00be7de226/pandas/tests/io/test_clipboard.py#L129): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

david-liu-brattle-1 committed:
Revert "Unit Test and whatsnew"

This reverts commit a8c098d596c39df7f4599bf1da516d00be7de226.

david-liu-brattle-1 committed:
Merge remote-tracking branch 'upstream/master' into fix-excel-clipboard

david-liu-brattle-1 committed:
Unit test for excel clipboard IO and updated whatsnew

david-liu-brattle-1 committed:
PEP8

chris-b1 commented on pandas/tests/io/test_clipboard.py:
```diff
@@ -123,6 +124,15 @@ def test_read_clipboard_infer_excel(self):
+        for dt in self.data_types:
+            for sep in ['\t', None]:
+                data = self.data[dt]
+                data.to_clipboard(excel=True, sep=sep)
```
Can you also test with `excel=False`?  In <=0.22 it would also be tab delimited

david-liu-brattle-1 replied:
You mean to test `data.to_clipboard(excel=False, sep=sep)`?  I don't think that was (or should be) tab delimited.  Do you mean `excel=None`?

chris-b1 replied:
Sorry, right, I meant to the default `df.to_clipboard()`  (sep=None, excel=None)

```
# pandas 0.22
In [72]: df = pd.DataFrame({'a': [1, 3], 'b': [4, 5]})

In [73]: df.to_clipboard()

# <paste>
In [74]: res = """	a	b
    ...: 0	1	4
    ...: 1	3	5"""

In [75]: res.count('\t')
Out[75]: 6

david-liu-brattle-1 replied:
Good point.  Should be added

david-liu-brattle-1 committed:
Test for function default values

david-liu-brattle-1 committed:
More robust clipboard tests

Added additional test df containing common delimiter symbols and quotes.

Added warning when attempting to copy excel format but an error is
caught

Default engine to "python" when reading clipboard with regex delimiter

david-liu-brattle-1 committed:
Test for correct shape when results aren't expected to exactly match

david-liu-brattle-1 committed:
PEP8

david-liu-brattle-1 committed:
Formatting

WillAyd commented on pandas/tests/io/test_clipboard.py:
```diff
@@ -69,18 +72,36 @@ def teardown_class(cls):
-        data.to_clipboard(excel=excel, sep=sep, encoding=encoding)
-        if sep is not None:
-            result = read_clipboard(sep=sep, index_col=0, encoding=encoding)
+        if excel in [None, True] and sep is not None and len(sep) > 1:
```
Admittedly not overly familiar with this code but having a hard time making sense of it - is this possible to parametrize? We typically avoid a lot of conditionals and loops in test scenarios

david-liu-brattle-1 replied:
Do you mean like this one? https://github.com/pandas-dev/pandas/blob/e033c0616158d3ba974456b4f84810492936b1fe/pandas/tests/io/test_common.py#L183-L188

WillAyd replied:
Yes that's one example. There's plenty of them in the tests so feel free to poke around

WillAyd commented on pandas/tests/io/test_clipboard.py:
```diff
@@ -124,6 +145,34 @@ def test_read_clipboard_infer_excel(self):
         tm.assert_frame_equal(res, exp)
 
+    def test_excel_clipboard_tabs(self):
+        for dt in self.data_types:
```
Same thing here - can we replace the loops with parametrization?

WillAyd commented on pandas/tests/io/test_clipboard.py:
```diff
@@ -60,6 +60,9 @@ def setup_class(cls):
         cls.data['utf8'] = pd.DataFrame({'a': ['µasd', 'Ωœ∑´'],
                                          'b': ['øπ∆˚¬', 'œ∑´®']})
+        # Test for quotes and common delimiters in text
+        cls.data['delim_symbols'] = pd.DataFrame({'a': ['"a,\t"b|c', 'd\tef´'],
```
What's the point of this?

david-liu-brattle-1 replied:
Excel can be very picky about how it handles copying/pasting text that includes delimiters or quote characters.  This dataframe caused some of the original clipboard tests to fail until I changed the to_clipboard and read_clipboard functions.

WillAyd commented on pandas/tests/io/test_clipboard.py:
```diff
@@ -124,6 +145,34 @@ def test_read_clipboard_infer_excel(self):
+                        # Expect tab delimited
+                        result = read_clipboard(sep='\t', index_col=0)
+                        tm.assert_frame_equal(data, result, check_dtype=False)
+                        assert clipboard_get().count('\t') > 0
```
Instead of counting the number of tabs can you construct the expected output and match exactly against that?

WillAyd commented on pandas/tests/io/test_clipboard.py:
```diff
@@ -69,18 +72,36 @@ def teardown_class(cls):
-        if sep is not None:
-            result = read_clipboard(sep=sep, index_col=0, encoding=encoding)
+        if excel in [None, True] and sep is not None and len(sep) > 1:
+            with tm.assert_produces_warning():
```
Typically for raising exceptions and warnings we split this off into separate tests

WillAyd commented on pandas/io/clipboards.py:
```diff
@@ -108,8 +112,9 @@ def to_clipboard(obj, excel=True, sep=None, **kwargs):  # pragma: no cover
-            pass
+        except TypeError:
+            warnings.warn('to_clipboard in excel mode requires a single \
+            character separator. Set "excel=false" or change the separator')
```
Capitalize `False`

jreback reviewed (changes requested):
yeah pls re-factor these tests to make it clear what is going on. ideally you can refactor in a commit, then do your changes in the following (or even a separate PR) to refactor first.

jreback commented:
@david-liu-brattle-1 can you rebase / update

david-liu-brattle-1 committed:
Merge branch 'master' into fix-excel-clipboard

david-liu-brattle-1 committed:
Rebase

david-liu-brattle-1 committed:
Typo

jorisvandenbossche commented on pandas/io/clipboards.py:
```diff
@@ -108,8 +112,9 @@ def to_clipboard(obj, excel=True, sep=None, **kwargs):  # pragma: no cover
-            pass
+        except TypeError:
+            warnings.warn('to_clipboard in excel mode requires a single \
+            character separator. Set "excel=False" or change the separator')
```
Can you use implicit line continuation instead of `\` ?

Like
```
            warnings.warn('to_clipboard in excel mode requires a single '
                          'character separator. Set "excel=False" or change '
                          'the separator')
```

jorisvandenbossche reviewed (commented):
You removed all the tests? (not fully clear if that was asked by the other reviewers) 

@WillAyd @chris-b1 can you take a new look?

jorisvandenbossche commented:
And the failing tests are actually clipboard tests that are failing, so should be addressed.

david-liu-brattle-1 commented:
@jorisvandenbossche Separate pr for the clipboard tests in #21163

jorisvandenbossche commented:
Ah, thanks, didn't see that! 
Only, the existing tests in this PR should still pass / be fixed for this to be merged.

david-liu-brattle-1 committed:
Fixed python 27 compatibility

david-liu-brattle-1 committed:
Merge branch 'master' into fix-excel-clipboard

david-liu-brattle-1 committed:
Style fix

Added comments to changes

Undo wrong file

Trying to fix mistakes

Attempt 2

Performance boost to printing DataFrames

PEP8 Changes

Revert "Merge remote-tracking branch 'refs/remotes/origin/fix-excel-clipboard'"

This reverts commit 8c7b0d00206db1125b8d83b2a2d4f028f8a13b18, reversing
changes made to 52cf5bcb389c3e486210709d7e31998bcfc81562.

Revert "Merge remote-tracking branch 'refs/remotes/origin/master' into fix-excel-clipboard"

This reverts commit cbb3999b95af2cab61b086b59972dedbe4aa5c42, reversing
changes made to a1c7422f67221363f63eac52f40868d5d2549020.

jreback commented on pandas/io/clipboards.py:
```diff
@@ -55,11 +56,17 @@ def read_clipboard(sep=r'\s+', **kwargs):  # pragma: no cover
-        sep = r'\t'
+        sep = '\t'
 
     if sep is None and kwargs.get('delim_whitespace') is None:
```
can you add some comments on these cases (including the existing ones).

jreback commented on pandas/io/clipboards.py:
```diff
@@ -55,11 +56,17 @@ def read_clipboard(sep=r'\s+', **kwargs):  # pragma: no cover
+    if sep == r'\s+' and kwargs.get('engine') is None:
+        kwargs['engine'] = 'python'
+
+    if kwargs.get('engine') == 'python' and compat.PY2:
```
why is the engine important here, shouldn't we alway encode in PY2?

jreback commented on pandas/io/clipboards.py:
```diff
@@ -108,8 +115,10 @@ def to_clipboard(obj, excel=True, sep=None, **kwargs):  # pragma: no cover
-            pass
+        except TypeError:
+            warnings.warn('to_clipboard in excel mode requires a single '
+                          'character separator. Set "excel=False" or change '
```
is this tested?

jreback commented on test (2).py:
```diff
@@ -0,0 +1,752 @@
+
+# The build cache system does string matching below this point.
+# if you change something, be careful.
+
```
looks like you copied in a random file, pls remove.

jreback commented on test (2).py:
```diff
@@ -0,0 +1,752 @@
+setup(name=DISTNAME,
+      maintainer=AUTHOR,
+      version=versioneer.get_version(),
+      packages=find_packages(include=['pandas', 'pandas.*']),
```
pls add tests

david-liu-brattle-1 committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-excel-clipboard

david-liu-brattle-1 committed:
Fix test

Fix Test

jorisvandenbossche reviewed (commented):
@david-liu-brattle-1 doesn't this need a test for the actual case that broke? (so the `excel=True` case)

david-liu-brattle-1 commented:
@jorisvandenbossche 
#21163 should have tests for all of the issues fixed in this PR.  I think that one will be merged first? In which case adding more tests here would be redundant right?

david-liu-brattle-1 committed:
Added warning for excel=False and sep!=None

jreback commented:
@david-liu-brattle-1 

- can you rebase on master
- add a whatsnew 0.23.2, IIRC this is a bug fix (IO section)
- remove the setup.py that was added.

ping on green.

jorisvandenbossche commented:
> #21163 should have tests for all of the issues fixed in this PR. I think that one will be merged first? In which case adding more tests here would be redundant right?

Ah, OK, the other PR is merged now, so then in this PR the `xfail`s should be removed?

david-liu-brattle-1 committed:
Merge branch 'master' into fix-excel-clipboard

david-liu-brattle-1 committed:
Removed xfail, add whatsnew

david-liu-brattle-1 committed:
Rebuild

david-liu-brattle-1 committed:
Typo fixes

david-liu-brattle-1 commented:
@jreback 
I think we're all set except for the setup.py file.  I'm not sure what exactly is happening with that, from my end it doesn't look like there are any changes to the existing file and I don't think I created that file.

chris-b1 commented:
Looks like file permissions were modified on `setup.py` (the `100755 → 100644` in the diff)

Should be able to just to reset to the current setup.py to back out the change `git checkout origin/master setup.py`

chris-b1 commented:
See also - https://stackoverflow.com/questions/1257592/how-do-i-remove-files-saying-old-mode-100755-new-mode-100644-from-unstaged-cha - might have some funky git settings

david-liu-brattle-1 committed:
permissions

jreback reviewed (approved):
rather minor point, other lgtm.

jreback commented:
@WillAyd merge when satisfied

WillAyd commented on pandas/io/clipboards.py:
```diff
@@ -55,11 +56,27 @@ def read_clipboard(sep=r'\s+', **kwargs):  # pragma: no cover
+    if len(sep) > 1 and kwargs.get('engine') is None:
+        kwargs['engine'] = 'python'
+    elif len(sep) > 1 and kwargs.get('engine') == 'c':
+        warnings.warn('from_clipboard with regex separator does not work'
```
What was the reasoning for going with a warning here instead of an error? Curious what actually happens if this comes up (question maybe applicable to other errors as well)

jorisvandenbossche replied:
I think that's a good question, we should check what `read_csv` does (warning or erroring) 
But anyhow, if we change it to an error, that would be for 0.24.0 IMO, so another PR.

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into david-liu-brattle-1-fix-excel-clipboard

jorisvandenbossche committed:
fix warning + small edits

jorisvandenbossche commented on pandas/io/clipboards.py:
```diff
@@ -121,14 +121,13 @@ def to_clipboard(obj, excel=True, sep=None, **kwargs):  # pragma: no cover
         except TypeError:
             warnings.warn('to_clipboard in excel mode requires a single '
-                          'character separator. Set "excel=False" or change '
-                          'the separator')
```
@david-liu-brattle-1 I removed the last sentence of this warning, as setting `excel=False` does not help as then you get the warning that the separator is ignored. It's simply that `to_clipboard` does not support multiple character separator at all.
